clar2wasm/src/words/sequences.rs

```diff
@@ -469,54 +469,55 @@ impl ComplexWord for Concat {
-        let ty = generator
+        let ret_ty = generator
             .get_expr_type(expr)
             .ok_or_else(|| GeneratorError::TypeError("concat expression must be typed".to_owned()))?
             .clone();
-        let (offset, _) = generator.create_call_stack_local(builder, &ty, false, true);
-
-        builder.local_get(offset);
-
-        // Traverse the lhs, leaving it on the data stack (offset, size)
-        let lhs = args.get_expr(0)?;
-        // WORKAROUND: typechecker issue for lists
-        generator.set_expr_type(lhs, ty.clone())?;
-        generator.traverse_expr(builder, lhs)?;
-
-        // Save the length of the lhs
-        let lhs_length = generator.module.locals.add(ValType::I32);
-        builder.local_tee(lhs_length);
+        let (ret_off, _) = generator.create_call_stack_local(builder, &ret_ty, false, true);
 
-        // Copy the lhs to the new sequence
-        builder.memory_copy(memory, memory);
+        let lhs_expr = args.get_expr(0)?;
+        let rhs_expr = args.get_expr(1)?;
 
-        // Load the adjusted destination offset
-        builder
-            .local_get(offset)
-            .local_get(lhs_length)
-            .binop(BinaryOp::I32Add);
+        let lhs_len = generator.module.locals.add(ValType::I32);
+        let rhs_len = generator.module.locals.add(ValType::I32);
 
-        // Traverse the rhs, leaving it on the data stack (offset, size)
-        let rhs = args.get_expr(1)?;
         // WORKAROUND: typechecker issue for lists
-        generator.set_expr_type(rhs, ty.clone())?;
-        generator.traverse_expr(builder, rhs)?;
-
-        // Save the length of the rhs
-        let rhs_length = generator.module.locals.add(ValType::I32);
-        builder.local_tee(rhs_length);
-
-        // Copy the rhs to the new sequence
+        generator.set_expr_type(lhs_expr, ret_ty.clone())?;
+        generator.set_expr_type(rhs_expr, ret_ty)?;
+
+        // Traverse the two arguments, setting up the stack to contain the
+        // arguments to the subsequent `memory.copy` instructions
+
+        /* [] */
+        builder.local_get(ret_off);
+        /* [ret_off] */
+        generator.traverse_expr(builder, lhs_expr)?;
+        builder.local_tee(lhs_len);
+        /* [ret_off, lhs_off, lhs_len] */
+        builder.local_get(ret_off);
+        /* [ret_off, lhs_off, lhs_len, ret_off] */
+        builder.local_get(lhs_len);
+        /* [ret_off, lhs_off, lhs_len, ret_off, lhs_len] */
+        builder.binop(BinaryOp::I32Add);
+        /* [ret_off, lhs_off, lhs_len, ret_off + lhs_len] */
+        generator.traverse_expr(builder, rhs_expr)?;
+        builder.local_tee(rhs_len);
+        /* [ret_off, lhs_off, lhs_len, ret_off + lhs_len, rhs_off, rhs_len] */
+
+        // Copy arguments to the return buffer. RHS is copied 1st, and LHS 2nd
+
+        /* [ret_off, lhs_off, lhs_len, ret_off + lhs_len, rhs_off, rhs_len] */
+        builder.memory_copy(memory, memory);
+        /* [ret_off, lhs_off, lhs_len] */
```

Comment:
You won't believe this, but there are so many comments that my brain can't parse the code :D
Might be just me though.

Reply:
Okay, I took a sheet of paper and rewrote this in Wat, it makes sense :)

Reply:
I actually do believe you, I used them as aid while rewriting. Would you prefer I remove them? Or perhaps I could leave them, but just on top?

Reply:
I tend to never put any, and I don't think I am a good example of documentation :sweat_smile: 
But yeah, maybe removing all the intermediary comments, regrouping all the `builder.x` into a single expression, and explaining on top what you are calculating could be more readable.
Not sure though, we should try.